Add automatic updating of deps.edn

plexus opened the pull request:
Use rewrite-clj to parse deps.edn and update dependencies without messing up
people's formatting.

Most of the functionality of depot.outdated.main has been pulled out to
prevent circular namespace dependencies. Main now only does the handling
of command line arguments.

plexus committed:
Add automatic updating of deps.edn

Use rewrite-clj to parse deps.edn and update dependencies without messing up
people's formatting.

Most of the functionality of depot.outdated.main has been pulled out to
prevent circular namespace dependencies. Main now only does the handling
of command line arguments.

Olical commented:
Nice! This is really cool and something that I thought would be requested at some point. Thank you very much, I'll take a good look through the code soon :smiley:

arichiardi commented:
Funny I was about to open an issue with this feature request :smile: 

I really like the `yarn` approach here:

https://asciinema.org/a/119515

plexus commented:
That is fancy indeed, but maybe something for another iteration.

@Olical let me know what you think, there are some things that could probably be better. Instead of `--update-all` I think it might be better to have separate `--update-user` and `--update-system` to deal with `~/.clojure/deps.edn` and `/usr/local/lib/clojure/deps.edn`

plexus commented on README.md:
```diff
@@ -10,7 +10,7 @@ You can try it out easily with this one liner:
 
 |          Dependency | Current | Latest |
-|---------------------+---------+--------|
+|---------------------|---------|--------|
```
Sorry about this, I think my Emacs does that automatically.

plexus commented on src/depot/outdated.clj:
```diff
@@ -0,0 +1,114 @@
+(ns depot.outdated
```
Everything in here is unmodified, just moved from `depot.outdated.main`

Olical commented:
Thanks for the comments, promise I'll get around to giving this a proper look over this week but I'm sure I'll be pretty much fine with it. No worries about the tables changing, just use Vim in future, thanks (:stuck_out_tongue_closed_eyes:)

If anyone seeing this is excited about this feature, please do point your `olical/depot` dependency at this repository and sha and let me know if it worked fine for your projects or not. It'll be a massive help!

arichiardi commented:
@Olical I am doing exactly that :dagger:

plexus committed:
Remove unused assert= and add docstrings.

plexus committed:
Restore formatting

plexus commented:
I added docstrings and removed some dead code.

If anyone wants to try this at home, then run this inside a directory containing a `deps.edn`

```
clojure -Sdeps '{:deps {olical/depot {:git/url "https://github.com/plexus/depot.git" :sha "6f6ec2d08de787b772ee5dc8270e582cb5e73627"}}}' -m depot.outdated.main --update 
```

Olical commented on src/depot/outdated.clj:
```diff
@@ -0,0 +1,114 @@
+        overrides (:override-deps args-map)
+        all-deps (merge (:deps deps-map) (:extra-deps args-map))]
+    (->> (for [[lib coord] all-deps
+               :let [_ (prn `(current-latest-map ~lib
```
Do you mean to prn here? It's pretty noisy in the output? If it's there to make the choices easier to understand I guess it makes more sense.

Olical replied:
Also worth noting that this kinda reintroduces an earlier problem where all of the maven output is interleaved with the actual useful information of what has been bumped. If no requests to maven are required then it's fine, I just see the clean version bump. All of the checks for newer versions just have side effects where they print out the "fetching ...." and URLs.

This was fixed before when the table output was added, so all of the changes are collected as the maven work is done and then printed out at the end. Maybe we could do something similar here? Or just log what versions changed and print them at the end instead of during the loop?

plexus replied:
Pretty sure this was just a debugging call that I forgot to take out. I removed it again.

Olical reviewed (changes requested):
Just some comments on stdout writing.

Olical commented on README.md:
```diff
@@ -32,6 +32,57 @@ $ clojure -Aoutdated -a outdated
+  rewrite-clj {:mvn/version "0.6.0"} -> {:mvn/version "0.6.1"}
+  cider/cider-nrepl {:mvn/version "0.17.0"} -> {:mvn/version "0.18.0"}
+  clj-time {:mvn/version "0.14.4"} -> {:mvn/version "0.15.1"}
+  olical/cljs-test-runner {:sha "5a18d41648d5c3a64632b5fec07734d32cca7671"} -> {:sha "da9710b389782d4637ef114176f6e741225e16f0"}
```
So with my comment about grouping the output, in this scenario you'd see a bunch of maven output between each line. At least, I do on my machine?

Olical commented on README.md:
```diff
@@ -41,7 +92,6 @@ This project is inspired by [lein-ancient][], it relies on [version-clj][] (by t
 Here's a few things I'd like to add some day, feel free to discuss or suggest more:
 
  * Searching for dependencies.
- * `deps.edn` manipulation like npm with `package.json`.
```
:tada: so cool to see one of these deleted

Olical commented on deps.edn:
```diff
@@ -1,11 +1,12 @@
-           :test {:extra-deps {clj-time {:mvn/version "0.14.4"}
+           :test {:extra-deps {clj-time {:mvn/version "0.15.1"}
                                olical/cljs-test-runner {:git/url "https://github.com/Olical/cljs-test-runner.git"
-                                                        :sha "5a18d41648d5c3a64632b5fec07734d32cca7671"}}}}}
```
So I left some of these outdated to test the functionality with :smile: maybe keep stuff inside the `:test` alias outdated and just checkout -- after running if you're testing.

plexus replied:
ok, I reverted them to their old versions.

Olical commented:
> That is fancy indeed, but maybe something for another iteration.
> 
> @Olical let me know what you think, there are some things that could probably be better. Instead of `--update-all` I think it might be better to have separate `--update-user` and `--update-system` to deal with `~/.clojure/deps.edn` and `/usr/local/lib/clojure/deps.edn`

I personally don't really touch my system and user level `deps.edn` files, I prefer to always keep things at the project level, I essentially pretend they don't exist. I worry I'll forget something is in a higher level file and it'll cause a repository to act weirdly for others. That's just me though! I'm happy having one flag for now, maybe split it if someone actually mentions it. Until then, YAGNI. Thanks for checking though!

plexus committed:
Remove debug prn call

plexus committed:
Leave some versions outdated for local testing.

plexus committed:
Get rid of --update-all, people can use --update with explicit paths.

plexus commented:
I removed `--update-all` again, I think it makes more sense to just have `--update`. People can do `--update ~/.clojure/deps.edn` explicitly if they want to update their user config, same for the system config.

Better to drop it for now then to drop it after people started using it.

Olical commented:
I gave it a test and it seems fine, worth noting that it requires Clojure 1.9+ because of `*print-namespace-maps*` (I mean, I think so, it failed for me on 1.8.0). Time to merge and deploy, thank you very much for your work!
